// json-serverless-lib/lib/storage/azure.fileshare.storage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AzureFileShareStorageAdapter = void 0;
var fs_1 = __importDefault(require("fs"));
var azure_fileshare_adapter_1 = require("./azure.fileshare.adapter");
var AzureFileShareStorageAdapter = /** @class */ (function () {
    function AzureFileShareStorageAdapter(azureStorageConnectionString, shareName, directoryName, jsonFilePath) {
        this.azureStorageConnectionString = azureStorageConnectionString;
        this.shareName = shareName;
        this.directoryName = directoryName;
        this.jsonFilePath = jsonFilePath;
    }
    AzureFileShareStorageAdapter.prototype.init = function () {
        console.log('initController');
        var db = JSON.parse(fs_1.default.readFileSync(this.jsonFilePath, 'UTF-8'));
        var storageAdapter = new azure_fileshare_adapter_1.AzureFileShareLowDBAdapter(db, this.azureStorageConnectionString, this.shareName, this.directoryName, this.jsonFilePath);
        return storageAdapter;
    };
    return AzureFileShareStorageAdapter;
}());
exports.AzureFileShareStorageAdapter = AzureFileShareStorageAdapter;
