// json-serverless-lib/lib/storage/dynamodb.storage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DynamoDBStorageAdapter = void 0;
var fs_1 = __importDefault(require("fs"));
var dynamodb_adapter_1 = require("./dynamodb.adapter");
var DynamoDBStorageAdapter = /** @class */ (function () {
    function DynamoDBStorageAdapter(table, keyId, jsonFilePath, region) {
        this.table = table;
        this.keyId = keyId;
        this.jsonFilePath = jsonFilePath;
        this.region = region;
    }
    DynamoDBStorageAdapter.prototype.init = function () {
        console.log('initController');
        var db = JSON.parse(fs_1.default.readFileSync(this.jsonFilePath, 'UTF-8'));
        var storageAdapter = new dynamodb_adapter_1.DynamoDBLowDBAdapter(db, this.table, this.keyId, this.region);
        return storageAdapter;
    };
    return DynamoDBStorageAdapter;
}());
exports.DynamoDBStorageAdapter = DynamoDBStorageAdapter;
