// json-serverless-lib/lib/app/app.config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogLevel = exports.AppRoutes = exports.AppConfig = void 0;
var AppConfig = /** @class */ (function () {
    function AppConfig() {
        this.readOnly = false;
        this.enableApiKeyAuth = false;
        this.jsonFile = 'db.json';
        this.enableJSONValidation = true;
        this.enableSwagger = true;
        this.logLevel = LogLevel.info;
        this.stackName = 'jsonsls';
        this.routes = new AppRoutes();
    }
    AppConfig.merge = function (t, u) { return Object.assign({}, t, u); };
    return AppConfig;
}());
exports.AppConfig = AppConfig;
var AppRoutes = /** @class */ (function () {
    function AppRoutes() {
        this.apiRoutePath = '/api';
        this.swaggerUIRoutePath = '/ui';
        this.swaggerSpecRoutePath = '/api-spec';
        this.graphqlRoutePath = '/graphql';
    }
    return AppRoutes;
}());
exports.AppRoutes = AppRoutes;
var LogLevel;
(function (LogLevel) {
    LogLevel["info"] = "info";
    LogLevel["debug"] = "debug";
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
