// json-serverless-lib/lib/specifications/swagger/swagger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Swagger = void 0;
var swaggerspec_1 = require("./swaggerspec");
var swagger_defgen_1 = require("./swagger.defgen");
var output_1 = require("../../utils/output");
var Swagger = /** @class */ (function () {
    function Swagger(server, config, basePath, apiRoutePath, packageJsonFilePath, specPath) {
        var _this = this;
        this.swaggerDefGen = new swagger_defgen_1.SwaggerDefGen();
        this.generateSpecification = function (json) {
            output_1.Output.setDebugInfo('Init Swagger');
            var swaggerSchemaDefinitions = _this.swaggerDefGen.generateDefinitions(json);
            var spec = _this.swaggerSpec.getSpec(_this.server, {}, _this.config.readOnly, _this.basePath, _this.apiRoutePath, _this.specPath);
            if (_this.config.enableApiKeyAuth) {
                var auth = {
                    type: 'apiKey',
                    in: 'header',
                    name: 'authorization',
                    description: 'All requests must include the `authorization` header containing your apikey.',
                };
                _this.swaggerSpec.addAuthentication(spec, auth);
            }
            _this.swaggerSpec.addSchemaDefitions(spec, swaggerSchemaDefinitions);
            return spec;
        };
        this.server = server;
        this.config = config;
        this.basePath = basePath;
        this.apiRoutePath = apiRoutePath;
        this.swaggerSpec = new swaggerspec_1.SwaggerSpec(packageJsonFilePath);
        this.specPath = specPath;
    }
    return Swagger;
}());
exports.Swagger = Swagger;
